cpp/src/arrow/ipc/metadata-internal.cc

```diff
@@ -176,6 +214,38 @@ static Status UnionToFlatBuffer(FBB& fbb, const std::shared_ptr<DataType>& type,
   return Status::OK();
 }
 
+static inline flatbuf::TimeUnit TimeUnitToFlatbuffer(TimeUnit unit) {
+  switch (unit) {
+    case TimeUnit::SECOND:
+      return flatbuf::TimeUnit_SECOND;
+    case TimeUnit::MILLI:
+      return flatbuf::TimeUnit_MILLISECOND;
+    case TimeUnit::MICRO:
+      return flatbuf::TimeUnit_MICROSECOND;
+    case TimeUnit::NANO:
+      return flatbuf::TimeUnit_NANOSECOND;
+  }
+
+  return flatbuf::TimeUnit_SECOND; // Default
+}
+
+static Offset TimestampToFlatbuffer(FBB& fbb, TimeUnit unit) {
+  return flatbuf::CreateTimestamp(fbb, TimeUnitToFlatbuffer(unit)).Union();
+}
+
+#define TIME_TO_FB(fbb, unit, type) \
+  flatbuf::Create ##  type(fbb, TimeUnitToFlatbuffer(unit)).Union();
+
+
+#define RETURN_BAD_TIME_UNIT(unit)                                  \
+  do {                                                              \
+        if ((unit < TimeUnit::MIN) || (unit > TimeUnit::MAX)) {     \
```

Comment:
Right, i.e. it should not be possible to return an invalid `TimeUnit` from a flatbuffer